pkg/agent/memberlist/cluster.go

```diff
@@ -0,0 +1,505 @@
+			}
+		}
+		consistentHashMap := newNodeConsistentHashMap()
+		consistentHashMap.Add(aliveAndMatchedNodes...)
+		c.consistentHashRWMutex.Lock()
+		defer c.consistentHashRWMutex.Unlock()
+		c.consistentHashMap[eip.Name] = consistentHashMap
+		c.notify(eip.Name)
+		return nil
+	}
+
+	if err := updateConsistentHash(eip); err != nil {
+		return err
+	}
+	return nil
+}
+
+func newNodeConsistentHashMap() *consistenthash.Map {
+	return consistenthash.New(defaultVirtualNodeReplicas, defaultHashFn)
+}
+
+func (c *Cluster) handleClusterNodeEvents(nodeEvent *memberlist.NodeEvent) {
+	node, event := nodeEvent.Node, nodeEvent.Event
+	switch event {
+	case memberlist.NodeJoin, memberlist.NodeLeave:
+		// When a Node joins cluster, all matched ExternalIPPools consistentHash should be updated.
+		// when a Node leave cluster, the Node may failed or deleted,
+		// if Node has been deleted, affected ExternalIPPool should enqueue, deleteNode handler has processed
+		// if the Node failed, ExternalIPPools consistentHash maybe changed, affected ExternalIPPool should enqueue.
+		coreNode, err := c.nodeLister.Get(node.Name)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				// Node has been deleted, deleteNode handler has processed.
+				klog.ErrorS(err, "Processing Node event, not found", "eventType", event)
+				return
+			}
+			klog.ErrorS(err, "Processing Node event, get Node failed", "eventType", event)
+			return
+		}
+		affectedEIPs := c.filterEIPsFromNodeLabels(coreNode)
+		c.enqueueExternalIPPools(affectedEIPs)
+		klog.InfoS("Processed Node event", "eventType", mapNodeEventType[event], "nodeName", node.Name, "affectedExternalIPPoolNum", len(affectedEIPs))
+	default:
+		klog.InfoS("Processed Node event", "eventType", mapNodeEventType[event], "nodeName", node.Name)
+	}
+}
+
+// aliveNodes returns the list of nodeNames in the cluster.
+func (c *Cluster) aliveNodes() sets.String {
+	nodes := sets.NewString()
+	for _, node := range c.mList.Members() {
+		nodes.Insert(node.Name)
+	}
+	return nodes
+}
+
+// ShouldSelectEgress the local Node in the cluster holds the same consistent hash ring for each ExternalIPPool,
+// when selecting an owner Node for an Egress, the local Node labels must match an ExternalIPPool nodeSelectors.
+// consistentHash.Get gets the closest item (Node name) in the hash to the provided key(egressName),
+// if the name of the local Node is equal to the name of the selected Node, returns true.
+func (c *Cluster) ShouldSelectEgress(egress *v1alpha2.Egress) (bool, error) {
+	eipName := egress.Spec.ExternalIPPool
+	if eipName == "" {
+		return false, nil
+	}
+	// Local Node must matches ExternalIPPool.
+	eip, err := c.externalIPPoolLister.Get(eipName)
+	if err != nil {
+		return false, fmt.Errorf("get ExternalIPPool %s failed: %+v", eipName, err)
+	}
+	localNode, err := c.nodeLister.Get(c.nodeName)
+	if err != nil {
+		return false, fmt.Errorf("get local Node %s error: %+v", c.nodeName, err)
+	}
+	if !c.matchNodeSelector(localNode, eip) {
```

Comment:
any comment?

Reply:
Perhaps there is little impact on improving performance I guess.

Reply:
Isn't `consistentHash.Get` more efficient than `matchNodeSelector`? The former is a binary search while the latter needs a few validation and conversion before creating a selector and the time spent on matching label may vary depending on the content of selector and label. I did a simple benchmark:
```
func BenchmarkMatchNodeSelector(b *testing.B) {
	node := &v1.Node{
		ObjectMeta: metav1.ObjectMeta{
			Labels: map[string]string{
				"foo": "bar",
			},
		},
	}
	pool := &crdv1a2.ExternalIPPool{
		Spec: crdv1a2.ExternalIPPoolSpec{
			NodeSelector: metav1.LabelSelector{
				MatchLabels: map[string]string{
					"foo": "bar",
				},
			},
		},
	}
	b.ReportAllocs()
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		matchNodeSelector(node, pool)
	}
}

func BenchmarkConsistentHashMapGet(b *testing.B) {
	m := newNodeConsistentHashMap()
	for i:=0;i<100;i++{
		m.Add(fmt.Sprintf("node%d",i))
	}
	b.ReportAllocs()
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		m.Get("abc")
	}
}
```
```
BenchmarkMatchNodeSelector-48             413698              2653 ns/op             441 B/op         11 allocs/op
BenchmarkConsistentHashMapGet-48         8163561               155.7 ns/op             8 B/op          1 allocs/op
```
Not to mention that the former needs to execute two Lister Get first. 

Not only the performance impact, my intention was to reduce redundant logic. Since we use a consistentHashMap to include all eligable Nodes and update it dynamically, no reason to check again whether a Node is eligable before using the map.

Reply:
It`s obviously consistentHash.Get more efficient than matchNodeSelector. But it needs to get consistentHashRWMutex firstly. Considering a large number of concurrent processesing of consistentHash updating events.